### ISSUE TITLE
Compiler should warn on `foreach (ref T2 t2 in enumThatIsRefT1)`

### ISSUE THREAD

**Issue by rickbrew** [dotnet/roslyn#61238]
`foreach (T value in enumerable)` is happy to insert an explicit cast to `T` from whatever type `enumerable` actually emits. IIRC this enabled things to work reasonably well back in the pre-generics age so you could `foreach (MyClass instance in arrayList)` because your `ArrayList`'s enumerator only gave out `object`s. Otherwise you'd have had to do `foreach (object instance in arrayList) { MyClass myClass = (MyClass)instance; ... }` _all over the place_. It was the least awful solution at the time, and also matched Java (iirc).

The thing is, looks like this still happens with `foreach (ref T2 value in enumerableOfRefT1s)` if `enumerableOfRefT1s` hands out an enumerator that does not implement `IEnumerable<T1>` but instead implements the duck-type pattern and has a `ref T1 Current` property.

I found this the hard way because a certain rendering path in Paint.NET was just not working; you'd draw with the Recolor Tool and it was not working; there was no effect on the canvas.

Turns out, it's because I had this code for the enumerator:

```cs
public unsafe readonly ref struct RegionRowPtr<T>
    where T : unmanaged
{
    private readonly T* ptr;
    private readonly int width;
    private readonly int y;
    ...
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public Enumerator GetEnumerator()
    {
        return new Enumerator(this.ptr, this.width);
    }
    ...
    public ref struct Enumerator
        //: IEnumerator<T>
    {
        private T* pCurrent;
        private int countLeft;
        ...
        public ref T Current
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => ref *this.pCurrent;
        }
        ...
    }
}
```    

And I was using it this way, elsewhere:

```cs
protected override unsafe void OnRenderContent(RegionPtr<ColorBgra> dstContent, Point2Int32 renderOffset)
{
    ColorBgra32 fillColor = this.Changes.FillColor;
    ColorBgra32 basisColor = this.basisColor;

    ThrowIfCancellationRequested();
    this.sampleSource.Render(dstContent, renderOffset);

    foreach (RegionRowPtr<ColorBgra> dstRow in dstContent.Rows)
    {
        ThrowIfCancellationRequested();

        // NOTE how dstRow's T is ColorBgra, but I'm foreach-ing with ColorBgra32
        foreach (ref ColorBgra32 dstColorRef in dstRow)
        {
            ColorBgra32 dstColor = dstColorRef;

            byte b = Int32Util.ClampToByte(dstColor.B + (fillColor.B - basisColor.B));
            byte g = Int32Util.ClampToByte(dstColor.G + (fillColor.G - basisColor.G));
            byte r = Int32Util.ClampToByte(dstColor.R + (fillColor.R - basisColor.R));
            byte a = Int32Util.ClampToByte(dstColor.A + (fillColor.A - basisColor.A));

            dstColorRef = ColorBgra32.FromBgra(b, g, r, a);
        }
    }
}
```

In my code, `PaintDotNet.ColorBgra` is the "old" color/pixel type, and `PaintDotNet.Imaging.ColorBgra32` is the new one, part of the newer more comprehensive imaging framework I've been working on. I'm still converting from the old to the new, little by little, over time. There's an implicit cast from `ColorBgra` to `ColorBgra32` (they are binary compatible). So instead of reading and writing to the refs that the enumerator was handing out, it was casting and then giving me `ref`s to the cast value (presumably on the stack). Reads were fine, writes went nowhere. Changing to `foreach (ref ColorBgra` fixed it.

I feel like this might be accidental? Surely the compiler should warn about this? I burned an hour on this, and thankfully it's the only location in my entire codebase where the text `foreach (ref ` exists, so it's not a widespread problem, but I could see it causing others a lot of pain and suffering.

**Comment by rickbrew** [comment]
BTW this also compiles just fine:

```cs
foreach (ref object dstColorRef in dstRow)
{
    ...
}
```

**Comment by 333fred** [comment]
> Note that the "ref locals and returns" feature lacks a specification.

This is why I asked Rick to open the bug here, rather than on csharplang, as I suspect this was never intended behavior. Maybe @VSadov knows whether this was intentional or was an oversight in implementation?

**Comment by jaredpar** [comment]
I think this is very likely just an oversight of implementation. There is no plausible way for `ref T Current` to work when the type of the `foreach` variable is different than `T`. It is feasible for the types to be different when dealing with `ref readonly` iterator variables but only when the iterator variable is a sub-type of `T`.

**Comment by RikkiGibson** [comment]
placed into a milestone based on my gut feeling here. feel free to adjust and assign out if desired.

**Comment by ArcadeMode** [comment]
Hi, 

I've looked into this issue and have found a potential solution that I would like to propose. 

When attempting to assign a ref indexer to a ref local the following errorcode may occur:
```csharp
var arr = new C2[] { new() };
ref C err2 = ref arr.AsSpan()[0]; //CS8173: The expression must be of type 'C' because it is being assigned by reference
class C { }
class C2 : C {}
```
I want to highlight that the site of the error is `arr.AsSpan()[0]`. In the aforementioned example by @RikkiGibson I would therefore expect the same error code to happen at site `arr.AsSpan()` (line 8). Does this sound like an appropriate solution to this issue?

If agreed I will open a PR, I have got the required changes ready in my fork. Looking forward to hearing back!

**Comment by RikkiGibson** [comment]
I agree with the basic idea here. Look for the enum member with value 8173 in ErrorCode.cs and start there. You should be able to find the code which is reporting `CS8173: The expression must be of type 'C'...`, find a reasonably corresponding place where the compiler converts a foreach variable to the element type, and simply add a similar check as what is used to report CS8173, when the foreach variable is a ref variable. Feel free to follow up here or in the [discord](https://discord.com/channels/143867839282020352/598678594750775301) if you have any questions.

**Comment by ArcadeMode** [comment]
Thanks for the reply, I have implemented just that! I will open the PR then 😄

**Comment by ArcadeMode** [comment]
I tried to join the discord but the link doesnt seem to work

**Comment by RikkiGibson** [comment]
See if this link works: https://discord.com/invite/tGJvv88

**Comment by 333fred** [comment]
> I tried to join the discord but the link doesnt seem to work

Which link? Which discord? There are multiple.

**Comment by ArcadeMode** [comment]
> See if this link works: https://discord.com/invite/tGJvv88

Thanks that one worked.

> > I tried to join the discord but the link doesnt seem to work
> 
> Which link? Which discord? There are multiple.

If there's any other one I should join to be able to reach out with questions then please invite me 😄

**Comment by 333fred** [comment]
> If there's any other one I should join to be able to reach out with questions then please invite me smile

Which link were you trying to follow that didn't work? The one in our README is correct.

**Comment by ArcadeMode** [comment]
> I agree with the basic idea here. Look for the enum member with value 8173 in ErrorCode.cs and start there. You should be able to find the code which is reporting `CS8173: The expression must be of type 'C'...`, find a reasonably corresponding place where the compiler converts a foreach variable to the element type, and simply add a similar check as what is used to report CS8173, when the foreach variable is a ref variable. Feel free to follow up here or in the [discord](https://discord.com/channels/143867839282020352/598678594750775301) if you have any questions.

I was refering to this link

**Comment by 333fred** [comment]
Ah. Yeah, the "right" invite is https://discord.gg/csharp.
